internal/controllers/bundledeployment/bundledeployment.go

```diff
@@ -198,53 +249,79 @@ func (c *controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
-		reason := rukpakv1alpha1.ReasonUnpackPending
-		status := metav1.ConditionTrue
-		message := fmt.Sprintf("Waiting for the %s Bundle to be unpacked", bundle.GetName())
-		if bundle.Status.Phase == rukpakv1alpha1.PhaseFailing {
-			reason = rukpakv1alpha1.ReasonUnpackFailed
-			status = metav1.ConditionFalse
-			message = fmt.Sprintf("Failed to unpack the %s Bundle", bundle.GetName())
-			if c := meta.FindStatusCondition(bundle.Status.Conditions, rukpakv1alpha1.TypeUnpacked); c != nil {
-				message = fmt.Sprintf("%s: %s", message, c.Message)
-			}
-		}
-		meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-			Type:    rukpakv1alpha1.TypeHasValidBundle,
-			Status:  status,
-			Reason:  reason,
-			Message: message,
-		})
+
+	if finalizerResult.Updated {
+		// The only thing outside the status that should ever change when handling finalizers
+		// is the list of finalizers in the object's metadata. In particular, we'd expect
+		// finalizers to be added or removed.
+		bd.ObjectMeta.Finalizers = finalizerBundleDelpoyment.ObjectMeta.Finalizers
+	}
+	if finalizerResult.StatusUpdated {
+		bd.Status = finalizerBundleDelpoyment.Status
+	}
+	if finalizerResult.Updated || finalizerResult.StatusUpdated || !bd.GetDeletionTimestamp().IsZero() {
 		return ctrl.Result{}, nil
 	}
 
-	meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-		Type:    rukpakv1alpha1.TypeHasValidBundle,
-		Status:  metav1.ConditionTrue,
-		Reason:  rukpakv1alpha1.ReasonUnpackSuccessful,
-		Message: fmt.Sprintf("Successfully unpacked the %s Bundle", bundle.GetName()),
-	})
+	unpackResult, err := c.unpacker.Unpack(ctx, bd)
+	if err != nil {
+		return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("source bundle content: %v", err))
+	}
+
+	switch unpackResult.State {
+	case unpackersource.StatePending:
+		updateStatusUnpackPending(&bd.Status, unpackResult)
+		// There must a limit to number of retries if status is stuck at
+		// unpack pending.
+		// Forcefully requing here, to ensure that the next reconcile
+		// is triggered.
+		// TODO: Caliberate the requeue interval if needed.
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	case unpackersource.StateUnpacking:
+		updateStatusUnpacking(&bd.Status, unpackResult)
+		return ctrl.Result{}, nil
+	case unpackersource.StateUnpacked:
+		storeFS, err := c.bundleDeploymentProcessor.Process(ctx, unpackResult.Bundle, bd)
+		if err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, err)
+		}
 
-	bundleFS, err := c.storage.Load(ctx, bundle)
+		if err := c.storage.Store(ctx, bd, storeFS); err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("persist bundle content: %v", err))
+		}
+		contentURL, err := c.storage.URLFor(ctx, bd)
+		if err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("get content URL: %v", err))
+		}
+		updateStatusUnpacked(&bd.Status, unpackResult, contentURL)
```

Comment:
I wonder if we should just store the unpacked contents directly without modification, and then move the logic that is in the processor into the existing handler?

Reply:
Modified in fd262e028de9b2bb81082408b1a7c12406dcaa56